File: cpp/Lec 2 Build-up Logical Thinking (Patterns)/pattern_06.cpp
// 12345
// 1234
// 123
// 12
// 1

//----------------------------------
// BRUTE FORCE APPROACH (Nested loops)
//----------------------------------
#include <iostream>
using namespace std;
int main()
{
    int n = 5;
    for (int i = n; i >= 1; i--)
    {
        for (int j = 1; j <= i; j++)
        {
            cout << j;
        }
        cout << endl;
    }
    return 0;
}

//----------------------------------
// BETTER APPROACH (String building)
//----------------------------------
#include <iostream>
#include <string>
using namespace std;
int main()
{
    int n = 5;
    for (int i = n; i >= 1; i--)
    {
        string s = "";
        for (int j = 1; j <= i; j++)
        {
            s += to_string(j);
        }
        cout << s << endl;
    }
    return 0;
}

//----------------------------------
// BEST APPROACH (Single pre-built string + substring)
//----------------------------------
#include <iostream>
#include <string>
using namespace std;
int main()
{
    int n = 5;
    const string s = "12345";
    for (int i = n; i >= 1; i--)
    {
        cout << s.substr(0, i) << '\n';
    }
    return 0;
}

File: python/Lec 2 Build-up Logical Thinking (Patterns)/pattern_06.py
# 12345
# 1234
# 123
# 12
# 1

#----------------------------------
# BRUTE FORCE APPROACH
#----------------------------------
n = 5
for i in range(n, 0, -1):
    line = ""
    for j in range(1, i + 1):
        line += str(j)  # O(n²) string copying
    print(line)

#----------------------------------
# BETTER APPROACH (Using join)
#----------------------------------
n = 5
for i in range(n, 0, -1):
    print(''.join(str(j) for j in range(1, i + 1)))

#----------------------------------
# BEST APPROACH (String slicing)
#----------------------------------
s = "12345"
print('\n'.join(s[:i] for i in range(len(s), 0, -1)))